minor-modes.lisp

```diff
@@ -0,0 +1,1083 @@
+                           (when scope-object
+                             (list scope-object))))
+                  (t (list (or scope-object
+                               (funcall (scope-current-object-function
+                                         (minor-mode-scope minor-mode))))))))))
+  (minor-mode-sync-keys-hook-function))
+
+(defun enable-minor-mode (minor-mode &optional scope-object)
+  "Enable MINOR-MODE. If MINOR-MODE is global, then enable it in all relevant
+objects. Otherwise enable it in the current object. If SCOPE-OBJECT is provided,
+use SCOPE-OBJECT instead of the current object, or include it in the list of
+current objects if MINOR-MODE is global"
+  (when (minor-mode-global-p minor-mode)
+    (pushnew minor-mode *active-global-minor-modes*))
+  (let ((run-hook nil))
+    (flet ((enable (object)
+             (cond ((typep object minor-mode)
+                    (restart-case 
+                        (error 'minor-mode-enable-error :mode minor-mode
+                                                        :object object
+                                                        :reason 'already-enabled)
+                      (continue () nil)))
+                   ((autoenable-minor-mode minor-mode object)
+                    (unless run-hook
+                      (setf run-hook object))))))
+      (mapc #'enable
+            (cond ((minor-mode-global-p minor-mode)
+                   (append (funcall (scope-all-objects-function
+                                     (minor-mode-scope minor-mode)))
+                           (when scope-object
+                             (list scope-object))))
+                  (t (list (or scope-object
+                               (funcall (scope-current-object-function
+                                         (minor-mode-scope minor-mode))))))))
+      (when run-hook
+        (run-hook-for-minor-mode #'minor-mode-hook minor-mode run-hook)
+        (run-hook-with-args *minor-mode-enable-hook* minor-mode run-hook))))
+  (minor-mode-sync-keys-hook-function))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Find Minor Modes ;;;
+;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defun sync-minor-modes (object)
+  "Sync the globally active minor modes in the object"
+  (loop for class in *active-global-minor-modes*
+        when (typep object (scope-type (minor-mode-scope class)))
+          do (autoenable-minor-mode class object)))
+
+(defun sync-all-minor-modes ()
+  "Loop through all recently created objects and ensure that the appropriate minor
+modes are enabled in them, then nullify the list of objects."
+  ;; This functions is needed because calling autoenable-minor-mode from within
+  ;; a method that accesses slots is implied to be undefined behavior, so we
+  ;; cant do this from within initialize-instance.
+  (let ((objects (prog1 (swm-class-new-objects (current-screen))
+                   (setf (swm-class-new-objects (current-screen)) nil))))
+    (when (and objects *active-global-minor-modes*)
```

Comment:
I'm unsure what you mean by `*active-global-minor-modes*` being disabled - it is intended to hold a list of all enabled minor modes that are also global, not neccessarily be a minor mode which is enabled/disable. The goal here is to get the list of new objects and then loop through all the enabled minor modes and enable them in objects where appropriate. If we have no active global minor modes then we dont need to do the loop, same if we have no new objects. Regardless we want to nullify the list of new objects.

Reply:
> I'm unsure what you mean by *active-global-minor-modes* being disabled
It was my bad, for some reason I was thinking it was a toggle instead of a list.

> Regardless we want to nullify the list of new objects.
Ok, then the code is correct as is. :+1: